Can not mix `include_cpp!` and `#[cxx::bridge]`

I was following the book and I can not get `include_cpp!` and `#[cxx::bridge]` to work together.
Am I doing something wrong? Or this is a bug?

## Expected Behavior

`cxxgen.h` contain both `include_cpp!` generation and `#[cxx::bridge]` generation.

## Actual Behavior

`cxxgen.h` contains only `#[cxx::bridge]` generation.

## Steps to Reproduce the Problem

1. These are my sources

Cargo.toml:
```toml
[package]
name = "autocxx-example"
version = "0.1.0"
edition = "2021"

[dependencies]
autocxx = "0.23.1"
cxx = "1.0.83"

[build-dependencies]
anyhow = "1.0.66"
autocxx-build = "0.23.1"
```

build.rs:
```rust
use anyhow::Result;

fn main() -> Result<()> {
    let path = std::path::PathBuf::from("src");
    let mut b = autocxx_build::Builder::new("src/main.rs", &[&path]).build()?;
    b.flag_if_supported("-std=c++14").file("src/deep.cc").compile("autocxx-demo");

    println!("cargo:rerun-if-changed=src/main.rs");
    println!("cargo:rerun-if-changed=src/deep.cc");
    println!("cargo:rerun-if-changed=src/deep.h");
    println!("cargo:rerun-if-changed=src/input.h");

    println!("cargo:rustc-link-lib=stdc++");

    Ok(())
}
```

src/main.rs:
```rust
use autocxx::extern_rust::extern_rust_function;
use autocxx::prelude::*;

include_cpp! {
    #include "deep.h"
    safety!(unsafe)
    generate!("DeepThought")
    extern_cpp_opaque_type!("Input", ffi2::Input)
}

#[cxx::bridge]
mod ffi2 {
    unsafe extern "C++" {
        include!("input.h");
        type Input;
    }
}

fn main() {
    println!("The answer to Life, The Universe and Everything is {}", ffi::DeepThought());
}

#[extern_rust_function]
pub fn call_me_from_cpp() {
    println!("Rust called from cpp");
}
```

src/deep.h:
```c++
#pragma once

#include <string>

std::string DeepThought();
```

src/deep.cc:
```c++
#include "deep.h"
#include "cxxgen.h"

std::string DeepThought() {
    call_me_from_cpp();
    return std::string("bla bla");
}
```

src/input.h:
```c++
#pragma once

struct Input {
    int val;
};
```

2. Run `cargo run`, the following error will be produced:

```
  cargo:warning=src/deep.cc: In function ‘std::string DeepThought()’:
  cargo:warning=src/deep.cc:5:5: error: ‘call_me_from_cpp’ was not declared in this scope
  cargo:warning=    5 |     call_me_from_cpp();
  cargo:warning=      |     ^~~~~~~~~~~~~~~~
  exit status: 1
  ```

3. `target/debug/build/autocxx-example-*/out/autocxx-build-dir/include/cxxgen.h` contains only code generation of `#[cxx::bridge]`

```c++
#pragma once
#include "input.h"

using Input = ::Input;
```

## Specifications

- Version:  autocxx 0.23.1, cxx 1.0.83
- Platform: Linux

--- a/engine/src/lib.rs
+++ b/engine/src/lib.rs
@@ -33,6 +33,7 @@ use parse_callbacks::AutocxxParseCallbacks;
 use parse_file::CppBuildable;
 use proc_macro2::TokenStream as TokenStream2;
 use regex::Regex;
+use std::cell::RefCell;
 use std::path::PathBuf;
 use std::rc::Rc;
 use std::{
@@ -704,7 +705,27 @@ pub struct CxxgenHeaderNamer<'a>(pub Box<dyn 'a + Fn() -> String>);
 
 impl Default for CxxgenHeaderNamer<'static> {
     fn default() -> Self {
-        Self(Box::new(|| "cxxgen.h".into()))
+        // The default implementation here is to name these headers
+        // cxxgen.h, cxxgen1.h, cxxgen2.h etc.
+        // These names are not especially predictable by callers and this
+        // behavior is not tested anywhere - so this is considered semi-
+        // supported, at best. This only comes into play in the rare case
+        // that you're generating bindings to multiple include_cpp!
+        // or a mix of include_cpp! and #[cxx::bridge] bindings.
+        let header_counter = Rc::new(RefCell::new(0));
+        Self(Box::new(move || {
+            let header_counter = header_counter.clone();
+            let header_counter_cell = header_counter.as_ref();
+            let mut header_counter = header_counter_cell.borrow_mut();
+            if *header_counter == 0 {
+                *header_counter += 1;
+                "cxxgen.h".into()
+            } else {
+                let count = *header_counter;
+                *header_counter += 1;
+                format!("cxxgen{count}.h")
+            }
+        }))
     }
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -499,11 +499,24 @@ pub mod extern_rust {
     /// generated `cxx` bindings, and this type will appear in the C++ header
     /// generated for use in C++.
     ///
+    /// # Finding these bindings from C++
+    ///
     /// You will likely need to forward-declare this type within your C++ headers
     /// before you can use it in such function signatures. autocxx can't generate
     /// headers (with this type definition) until it's parsed your header files;
     /// logically therefore if your header files mention one of these types
     /// it's impossible for them to see the definition of the type.
+    ///
+    /// If you're using multiple sets of `include_cpp!` directives, or
+    /// a mixture of `include_cpp!` and `#[cxx::bridge]` bindings, then you
+    /// may be able to `#include "cxxgen.h"` to refer to the generated C++
+    /// function prototypes. In this particular circumstance, you'll want to know
+    /// how exactly the `cxxgen.h` header is named, because one will be
+    /// generated for each of the sets of bindings encountered. The pattern
+    /// can be set manually using `autocxxgen`'s command-line options. If you're
+    /// using `autocxx`'s `build.rs` support, those headers will be named
+    /// `cxxgen.h`, `cxxgen1.h`, `cxxgen2.h` according to the order in which
+    /// the `include_cpp` or `cxx::bridge` bindings are encountered.
     pub use autocxx_macro::extern_rust_type;
 
     /// Declare that a given function is a Rust function which is to be exported
@@ -513,6 +526,9 @@ pub mod extern_rust {
     /// #[extern_rust_function]
     /// pub fn call_me_from_cpp() { }
     /// ```
+    ///
+    /// See [`extern_rust_type`] for details of how to find the generated
+    /// declarations from C++.
     pub use autocxx_macro::extern_rust_function;
 }
 

